### DISCUSSION TITLE
Problems when using `imports`

### DISCUSSION THREAD

**Discussion by FinYang** [ropensci/targets#945] [Help]
Hi there, 

I only started using `targets` yesterday. It significantly improved my workflow (Thank you for such a great package), but I have been having trouble using `imports`. I am developing package(s) while using them in a pipeline. This is the same use case described in https://books.ropensci.org/targets/packages.html#package-based-invalidation. 

### It does not work with "Imports" and "Suggests":

 `dplyr` "Imports" `tidyselect`, and `tidyselect` "Suggests" `dplyr`
``` r
library(targets)
tar_script({
    tar_option_set(packages = c("dplyr", "tidyselect"), imports = c("dplyr", 
        "tidyselect"))
    list(tar_target(x, 1))
})
tar_make()
#> Error:
#> ! Error running targets::tar_make()
#>   Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
#>   Tips: https://books.ropensci.org/targets/debugging.html
#>   Last error: igraph::is_dag(graph) is not TRUE


#> Backtrace:
#>     ▆
#>  1. └─targets::tar_make()
#>  2.   └─targets:::callr_outer(...)
#>  3.     ├─targets:::if_any(...)
#>  4.     └─targets:::callr_error(condition = out, fun = fun)
#>  5.       └─targets::tar_throw_run(message)
#>  6.         └─targets::tar_error(...)
#>  7.           └─rlang::abort(message = message, class = class, call = tar_empty_envir)
```

<sup>Created on 2022-10-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### or "Depends":
`fable` “Depends" `fabletools`

``` r
library(targets)
tar_script({
    tar_option_set(packages = c("fable", "fabletools"), imports = c("fable", 
        "fabletools"))
    list(tar_target(x, 1))
})
tar_make()
#> Error:
#> ! Error running targets::tar_make()
#>   Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
#>   Tips: https://books.ropensci.org/targets/debugging.html
#>   Last error: igraph::is_dag(graph) is not TRUE

#> Backtrace:
#>     ▆
#>  1. └─targets::tar_make()
#>  2.   └─targets:::callr_outer(...)
#>  3.     ├─targets:::if_any(...)
#>  4.     └─targets:::callr_error(condition = out, fun = fun)
#>  5.       └─targets::tar_throw_run(message)
#>  6.         └─targets::tar_error(...)
#>  7.           └─rlang::abort(message = message, class = class, call = tar_empty_envir)
```

<sup>Created on 2022-10-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### I cannot trigger any rerun if I only modify S3 methods
(which is kinda expected but still)

Code generating reprex:
```r
targets::tar_reprex(
  pipeline = {
    tar_option_set(
      packages = "targetsReprex",
      imports = "targetsReprex")
    list(tar_target(output_s3, test_s3(1)) )
  },
  run = {
    remotes::install_github(
      "FinYang/targetsReprex",
      ref = "485c2f8ee865bd5c588213058c7b8b50a2400db0",
      quiet = TRUE,
      force = TRUE)
    tar_make()
    digest::digest(tar_read(output_s3))

    # test_s3.numeric method changed
    # but not the generic test_s3
    remotes::install_github(
      "FinYang/targetsReprex",
      ref = "240fa544e8c781b20131a2a4478c14ef190d3a43",
      quiet = TRUE,
      force = TRUE)
    tar_make()
    digest::digest(tar_read(output_s3))
    # It will not invalidate the target


    # test_s3 generic changes
    remotes::install_github(
      "FinYang/targetsReprex",
      ref = "594e837683249f7b698009119ae7c465f7a01b5f",
      quiet = TRUE,
      force = TRUE)
    tar_make()
    digest::digest(tar_read(output_s3))
  }
)
```

Reprex:
``` r
library(targets)
tar_script({
    tar_option_set(packages = "targetsReprex", imports = "targetsReprex")
    list(tar_target(output_s3, test_s3(1)))
})
remotes::install_github("FinYang/targetsReprex", ref = "485c2f8ee865bd5c588213058c7b8b50a2400db0", 
    quiet = TRUE, force = TRUE)
tar_make()
#> • start target output_s3
#> • built target output_s3 [0 seconds]
#> • end pipeline [0.05 seconds]
digest::digest(tar_read(output_s3))
#> [1] "db8e490a925a60e62212cefc7674ca02"
remotes::install_github("FinYang/targetsReprex", ref = "240fa544e8c781b20131a2a4478c14ef190d3a43", 
    quiet = TRUE, force = TRUE)
tar_make()
#> ✔ skip target output_s3
#> ✔ skip pipeline [0.03 seconds]
digest::digest(tar_read(output_s3))
#> [1] "db8e490a925a60e62212cefc7674ca02"
remotes::install_github("FinYang/targetsReprex", ref = "594e837683249f7b698009119ae7c465f7a01b5f", 
    quiet = TRUE, force = TRUE)
tar_make()
#> • start target output_s3
#> • built target output_s3 [0 seconds]
#> • end pipeline [0.05 seconds]
digest::digest(tar_read(output_s3))
#> [1] "e5b57f323c7b3719bbaaf9f96b260d39"
```

<sup>Created on 2022-10-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

## 3. List of functions in the package gets invalidated every time even if there is no change

This is a bit tricky to write out a reprex (I really tried) because it involves launching a separate R session. My current workflow is 
1. Modify the package
2. Build and install
3. `tar_make` in a different R session

The problem is when I build and install, any object in the package that is not a function but contains a function (or environment?) gets invalidated, even if there is no change in that object. A simple example would be 
```
list_fun <- list(function(){})
```
I suspect it has something to do with how an environment is generated when a package is installed, but I don't know enough about that process to figure it out.

## Overall
This means while most of my functions are in different packages, I cannot track more than one package at once, I cannot trigger reruns of the parts I modify (S3 method), and the whole other half of the pipeline gets invalidated unexpectedly with "list of functions" which can take a day or two to rerun if I add more targets.

I tried to come up with a workaround using `devtools::load_all` to get all package dependencies and `tar_source` to track objects in the package, but it just gets messier. Right now I just manually invalidate parts that I modify (or I might try still using `imports` and fixing some of the targets with `tar_cue`). In general, I have benefited a lot from the data storage of `targets` (so I don't have to manually read and write files) and the visualisation and monitoring of the pipeline, not so much from change tracking.

Edit:
Just discovered another (potential) problem with `imports`: tracked packages have different hashes if installed on different computers. This makes it impossible to use some online file hosting service (I'm using OneDrive) to manage data and `tar_make` on different devices without triggering reruns on each device.

**Comment by wlandau** [comment]
1. The `imports` option is intended just for the analysis packages you develop yourself, whose functions you are likely to change yourself. It is not a great fit for infrastructure packages like `dplyr` or `tidyverse`, and it is not aware of the package dependencies (`Imports:`, `Depends:`, `Suggests:`) in the `DESCRIPTION` files of those packages.
2. `targets` learns the dependency structure of the pipeline using static code analysis, which means it reads your code without running it. This is an inherent limitation. It does not know what to do with S3 because S3 dispatch happens at runtime while the pipeline is already configured and running.
3. This is another limitation and a consequence of the design philosophy of `targets`, and it is related to several of the issues referenced from https://github.com/ropensci/drake/issues/345. To get around this, you could avoid creating targets that depend on that object whose hash keeps changing, or you could create an intermediate target that just takes out the important and less complicated parts of that object.

These are limitations that I can't make any easier in the package itself. But you could compromise and track e.g. the output of `packageDescription()` as a target and then make it a dependency of downstream targets. Similarly, if there are especially important data objects, you could assign those to `tar_option_get("envir")` and make them dependencies of the appropriate targets.
